src/pubtools/sign/signers/cosignsigner.py

```diff
@@ -0,0 +1,291 @@
+        """Return list of supported operations."""
+        return self.SUPPORTED_OPERATIONS
+
+    def sign(self: CosignSigner, operation: SignOperation) -> SigningResults:
+        """Run signing operation.
+
+        :param operation: signing operation
+        :type operation: SignOperation
+
+        :return: SigningResults
+        """
+        if isinstance(operation, ContainerSignOperation):
+            return self.container_sign(operation)
+        else:
+            raise UnsupportedOperation(operation)
+
+    def container_sign(self: CosignSigner, operation: ContainerSignOperation):
+        """Run container signing operation.
+
+        :param operation: signing operation
+        :type operation: ContainerSignOperation
+
+        :return: SigningResults
+        """
+        set_log_level(LOG, self.log_level)
+        if operation.references and len(operation.digests) != len(operation.references):
+            raise ValueError("Digests must pairs with references")
+
+        signer_results = CosignSignerResults(status="ok", error_message="")
+
+        operation_result = ContainerSignResult(signing_key=operation.signing_key, signed_claims=[])
+
+        signing_results = SigningResults(
+            signer=self,
+            operation=operation,
+            signer_results=signer_results,
+            operation_result=operation_result,
+        )
+
+        outputs = {}
+        processes = {}
+        common_args = [
+            self.cosign_bin,
+            "-t",
+            self.timeout,
+            "sign",
+            "--key",
+            operation.signing_key,
+            "--allow-http-registry=%s" % ("true" if self.allow_http_registry else "false"),
+            "--allow-insecure-registry=%s" % ("true" if self.allow_insecure_registry else "false"),
+            "--rekor-url",
+            self.rekor_url,
+            "--tlog-upload=%s" % ("true" if self.upload_tlog else "false"),
+        ]
+        if operation.references:
+            for ref, digest in zip(operation.references, operation.digests):
+                repo, tag = ref.rsplit(":", 1)
+                processes[f"{repo}:{digest}"] = run_command(
+                    common_args + ["-a", f"tag={tag}", f"{repo}@{digest}"],
+                    env=self.env_variables,
+                )
+        else:
+            for ref_digest in operation.digests:
+                processes[f"{ref_digest}"] = run_command(
+                    common_args + [ref_digest], env=self.env_variables
+                )
+        for ref, process in processes.items():
+            stdout, stderr = process.communicate()
+            outputs[ref] = (stdout, stderr, process.returncode)
+
+        operation_result = ContainerSignResult(signing_key=operation.signing_key, signed_claims=[])
+
+        for ref, (stdout, stderr, returncode) in outputs.items():
+            if returncode != 0:
+                operation_result.signed_claims.append(stderr)
```

Comment:
Is this a mistake?

Reply:
The reason why I am confused is that there is a structure CosignSignerResults with error_message field, which is totally unused. Instead you (inexplicably) put the error message to signed_claims.

Reply:
I changed the structure to make it less confusing